app/services/fact_importer.rb

```diff
@@ -86,8 +85,13 @@ def add_new_facts
     logger.debug("Merging facts for '#{host}': added #{@counters[:added]} facts")
   end
 
+  def preload_fact_names
+    # :type is needed because custom facts usually inherits from FactName so they would be included in the list
+    fact_name_class.where(:type => fact_name_class).group(:name).maximum(:id)
```

Comment:
Why are you doing group here? Since `name` is unique across types, you can simply do:

    Hash[fact_name_class.where(:type => fact_name_class).pluck(:name, :id)]

which will definitely be faster.